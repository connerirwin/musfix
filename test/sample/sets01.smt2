; Convert Set to Set, parse Set Sort and Set Literal properly
; List are some constructed data type
(constraint
  (forall ((v Int) (m1 (Set Int)))
    (=> (= m1 {})
        (not (in 100 m1)))))
; In the negative version, the not has been removed 

(constraint
  (forall ((v Int) (m1 (Set Int)) (m2 (Set Int)))
    (=> (&& (= m1 {})
          (= m2 (union (union m1 (Set 10)) (Set 20))))
        (not (in 100 m2)))))

(constraint
  (forall ((v Int) (m1 (Set Int)) (m2 (Set Int)))
    (=> (&& (= m1 {})
          (= m2 (union (union m1 (Set 10)) (Set 20))))
        (in 10 m2))))

(constraint
  (forall ((v Int) (m1 (Set Int)) (m2 (Set Int)) (m3 (Set Int)))
    (=> (&& (&& (= m1 {})
          (= m2 (union (union m1 (Set 10)) (Set 20))))
          (= m3 (union (union m1 (Set 20)) (Set 10))))
        (= m2 m3))))

(constraint
  (forall ((v Int) (m1 (Set Int)) (m2 (Set Int)) (m3 (Set Int))
            (m4 (Set Int)) (m5 (Set Int)))
    (=> (&& (&& (&& (&& (= m1 {})
          (= m2 (union (union m1 (Set 10)) (Set 20))))
          (= m3 (union (union m1 (Set 20)) (Set 10))))
          (= m4 (union m1 (Set 10))))
          (= m5 (union m1 (Set 20))))
        (= m2 (union m4 m5)))))

; (constraint
;   (forall ()
;     (=> True False)))
