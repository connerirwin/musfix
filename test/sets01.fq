// constant Set_empty : (func(1, [int; Set_Set @(0)]))
constant Set_sng : (func(1, [@(0); (Set_Set  @(0))]))

bind 1 m1 : {v : Set_Set Int | v = Set_empty 0}
bind 2 m2 : {v : Set_Set Int | v = (Set_cup (Set_cup m1 (Set_sng 10)) (Set_sng 20)) }
bind 3 m3 : {v : Set_Set Int | v = (Set_cup (Set_cup m1 (Set_sng 20)) (Set_sng 10)) }
bind 4 m4 : {v : Set_Set Int | v = (Set_cup m1 (Set_sng 10)) }
bind 5 m5 : {v : Set_Set Int | v = (Set_cup m1 (Set_sng 20)) }

constraint:
  env [ 1 ]
  lhs {v : int | true }
  rhs {v : int | not (Set_mem 100 m1)  }
  id 1 tag []

constraint:
  env [ 1; 2 ]
  lhs {v : int | true }
  rhs {v : int | not (Set_mem 100 m2) }
  id 2 tag []

constraint:
  env [ 1; 2 ]
  lhs {v : int | true }
  rhs {v : int | Set_mem 10 m2 }
  id 3 tag []

constraint:
  env [ 1; 2; 3 ]
  lhs {v : int | true }
  rhs {v : int | m2 = m3 }
  id 4 tag []

constraint:
  env [ 1; 2; 3; 4; 5 ]
  lhs {v : int | true }
  rhs {v : int | m2 = Set_cup m4 m5 }
  id 5 tag []
