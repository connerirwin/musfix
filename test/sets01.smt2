; Convert Set_t to Set, parse Set Sort and Set Literal properly
(constraint
  (forall ((v Int) (m1 (Set_t Int)))
    (=> (= m1 [])
        (not (in 100 m1)))))

(constraint
  (forall ((v Int) (m1 (Set_t Int)) (m2 (Set_t Int)))
    (=> (&& (= m1 [])
          (= m2 (union (union m1 (Set 10)) (Set 20))))
        (not (in 100 m2)))))

(constraint
  (forall ((v Int) (m1 (Set_t Int)) (m2 (Set_t Int)))
    (=> (&& (= m1 [])
          (= m2 (union (union m1 (Set 10)) (Set 20))))
        (in 10 m2))))

(constraint
  (forall ((v Int) (m1 (Set_t Int)) (m2 (Set_t Int)) (m3 (Set_t Int)))
    (=> (&& (&& (= m1 [])
          (= m2 (union (union m1 (Set 10)) (Set 20))))
          (= m3 (union (union m1 (Set 20)) (Set 10))))
        (= m2 m3))))

(constraint
  (forall ((v Int) (m1 (Set_t Int)) (m2 (Set_t Int)) (m3 (Set_t Int))
            (m4 (Set_t Int)) (m5 (Set_t Int)))
    (=> (&& (&& (&& (&& (= m1 [])
          (= m2 (union (union m1 (Set 10)) (Set 20))))
          (= m3 (union (union m1 (Set 20)) (Set 10))))
          (= m4 (union m1 (Set 10))))
          (= m5 (union m1 (Set 20))))
        (= m2 m3))))
