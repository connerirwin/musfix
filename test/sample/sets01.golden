
Preparing to find greatest fixpoint...

Inputs
--------
Reading from file test/sample/sets01.msmt
HornConstraint [Var IntS "v",Var (SetS IntS) "m1"] (Binary Implies (Binary Eq (Var (SetS IntS) "m1") (SetLit IntS [])) (Unary Not (Binary Member (IntLit 100) (Var (SetS IntS) "m1"))))
HornConstraint [Var IntS "v",Var (SetS IntS) "m1",Var (SetS IntS) "m2"] (Binary Implies (Binary And (Binary Eq (Var (SetS IntS) "m1") (SetLit IntS [])) (Binary Eq (Var (SetS IntS) "m2") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 10])) (SetLit IntS [IntLit 20])))) (Unary Not (Binary Member (IntLit 100) (Var (SetS IntS) "m2"))))
HornConstraint [Var IntS "v",Var (SetS IntS) "m1",Var (SetS IntS) "m2"] (Binary Implies (Binary And (Binary Eq (Var (SetS IntS) "m1") (SetLit IntS [])) (Binary Eq (Var (SetS IntS) "m2") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 10])) (SetLit IntS [IntLit 20])))) (Binary Member (IntLit 10) (Var (SetS IntS) "m2")))
HornConstraint [Var IntS "v",Var (SetS IntS) "m1",Var (SetS IntS) "m2",Var (SetS IntS) "m3"] (Binary Implies (Binary And (Binary And (Binary Eq (Var (SetS IntS) "m1") (SetLit IntS [])) (Binary Eq (Var (SetS IntS) "m2") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 10])) (SetLit IntS [IntLit 20])))) (Binary Eq (Var (SetS IntS) "m3") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 20])) (SetLit IntS [IntLit 10])))) (Binary Eq (Var (SetS IntS) "m2") (Var (SetS IntS) "m3")))
HornConstraint [Var IntS "v",Var (SetS IntS) "m1",Var (SetS IntS) "m2",Var (SetS IntS) "m3",Var (SetS IntS) "m4",Var (SetS IntS) "m5"] (Binary Implies (Binary And (Binary And (Binary And (Binary And (Binary Eq (Var (SetS IntS) "m1") (SetLit IntS [])) (Binary Eq (Var (SetS IntS) "m2") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 10])) (SetLit IntS [IntLit 20])))) (Binary Eq (Var (SetS IntS) "m3") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 20])) (SetLit IntS [IntLit 10])))) (Binary Eq (Var (SetS IntS) "m4") (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 10])))) (Binary Eq (Var (SetS IntS) "m5") (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 20])))) (Binary Eq (Var (SetS IntS) "m2") (Binary Union (Var (SetS IntS) "m4") (Var (SetS IntS) "m5"))))

QMAP
--------
fromList []

Candidates
--------
[Candidate {solution = fromList [], validConstraints = fromList [Binary Implies (Binary Eq (Var (SetS IntS) "m1") (SetLit IntS [])) (Unary Not (Binary Member (IntLit 100) (Var (SetS IntS) "m1"))),Binary Implies (Binary And (Binary Eq (Var (SetS IntS) "m1") (SetLit IntS [])) (Binary Eq (Var (SetS IntS) "m2") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 10])) (SetLit IntS [IntLit 20])))) (Unary Not (Binary Member (IntLit 100) (Var (SetS IntS) "m2"))),Binary Implies (Binary And (Binary Eq (Var (SetS IntS) "m1") (SetLit IntS [])) (Binary Eq (Var (SetS IntS) "m2") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 10])) (SetLit IntS [IntLit 20])))) (Binary Member (IntLit 10) (Var (SetS IntS) "m2")),Binary Implies (Binary And (Binary And (Binary Eq (Var (SetS IntS) "m1") (SetLit IntS [])) (Binary Eq (Var (SetS IntS) "m2") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 10])) (SetLit IntS [IntLit 20])))) (Binary Eq (Var (SetS IntS) "m3") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 20])) (SetLit IntS [IntLit 10])))) (Binary Eq (Var (SetS IntS) "m2") (Var (SetS IntS) "m3")),Binary Implies (Binary And (Binary And (Binary And (Binary And (Binary Eq (Var (SetS IntS) "m1") (SetLit IntS [])) (Binary Eq (Var (SetS IntS) "m2") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 10])) (SetLit IntS [IntLit 20])))) (Binary Eq (Var (SetS IntS) "m3") (Binary Union (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 20])) (SetLit IntS [IntLit 10])))) (Binary Eq (Var (SetS IntS) "m4") (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 10])))) (Binary Eq (Var (SetS IntS) "m5") (Binary Union (Var (SetS IntS) "m1") (SetLit IntS [IntLit 20])))) (Binary Eq (Var (SetS IntS) "m2") (Binary Union (Var (SetS IntS) "m4") (Var (SetS IntS) "m5")))], invalidConstraints = fromList [], label = "0"}]


Final candidates: 
0: [] (5 0)

